ludwig/data/dataframe/dask.py

```diff
@@ -47,13 +66,45 @@ def set_parallelism(self, parallelism):
         self._parallelism = parallelism
 
     def df_like(self, df: dd.DataFrame, proc_cols: Dict[str, dd.Series]):
+        """Outer joins the given DataFrame with the given processed columns.
+
+        NOTE: If any of the processed columns have been repartitioned, the original index is replaced with a
+        monotonically increasing index, which is used to define the new divisions and align the various partitions.
+        """
         # Our goal is to preserve the index of the input dataframe but to drop
         # all its columns. Because to_frame() creates a column from the index,
         # we need to drop it immediately following creation.
         dataset = df.index.to_frame(name=TMP_COLUMN).drop(columns=TMP_COLUMN)
+
+        repartitioned_cols = {}
         for k, v in proc_cols.items():
-            v.divisions = dataset.divisions
-            dataset[k] = v
+            if v.npartitions == dataset.npartitions:
+                # Outer join cols with equal partitions
+                v.divisions = dataset.divisions
+                dataset[k] = v
+            else:
+                # If partitions have changed (e.g. due to conversion from Ray dataset), we handle separately
+                repartitioned_cols[k] = v
+
+        if repartitioned_cols:
+            if not dataset.known_divisions:
```

Comment:
Is this primarily to handle the image columns, which lack a global index? I just want to make sure we're handling things like numerical columns correctly, where we rely on the index in the case where we drop missing values from the column.

Reply:
Great question– I left a hopefully descriptive (if long) comment in `preprocessing.build_dataset` (at time of writing this comment, line 1087). Essentially, we only need a globally unique index if we know there will be a repartition step (i.e. with image and audio features). If there will NOT be a repartition step (i.e. purely tabular data), partition-level unique indices are sufficient because we don't need to do a global re-alignment of partitions– the join op in `df_like` can happen at the partition-level as per usual.